src/scc.js

```diff
@@ -0,0 +1,91 @@
+import calculateScc from '@rtsao/scc';
+import { hashObject } from 'eslint-module-utils/hash';
+import resolve from 'eslint-module-utils/resolve';
+import ExportMapBuilder from './exportMap/builder';
+import childContext from './exportMap/childContext';
+
+let cache = new Map();
+let exportMapBuilder = ExportMapBuilder;
+
+export default class StronglyConnectedComponentsBuilder {
+  static injectExportMapBuilder(value) {
+    exportMapBuilder = value;
+  }
+
+  static resetExportMapBuilder() {
+    exportMapBuilder = ExportMapBuilder;
+  }
+
+  static clearCache() {
+    cache = new Map();
+  }
+
+  static get(source, context) {
+    const path = resolve(source, context);
+    if (path == null) { return null; }
+    return StronglyConnectedComponentsBuilder.for(childContext(path, context));
+  }
+
+  static for(context) {
+    const cacheKey = context.cacheKey || hashObject(context).digest('hex');
+    if (cache.has(cacheKey)) {
+      return cache.get(cacheKey);
+    }
+    const scc = StronglyConnectedComponentsBuilder.calculate(context);
+    cache.set(cacheKey, scc);
+    return scc;
+  }
+
+  static calculate(context) {
+    const exportMap = exportMapBuilder.for(context);
+    const adjacencyList = this.exportMapToAdjacencyList(exportMap);
+    const calculatedScc = calculateScc(adjacencyList);
+    return StronglyConnectedComponentsBuilder.calculatedSccToPlainObject(calculatedScc);
+  }
+
+  /** @returns {Map<string, Set<string>>} for each dep, what are its direct deps */
+  static exportMapToAdjacencyList(initialExportMap) {
+    const adjacencyList = new Map();
+    // BFS
+    function visitNode(exportMap) {
+      if (!exportMap) {
+        return;
+      }
+      exportMap.imports.forEach((v, importedPath) => {
+        const from = exportMap.path;
+        const to = importedPath;
+
+        if (!adjacencyList.has(from)) {
+          adjacencyList.set(from, new Set());
```

Comment:
would it be faster to defer creation of these new sets until they're needed?

Reply:
They are deferred already.
`exportMapToAdjacencyList` is called from `calculate` which is called from `for` which is called from `get`, which is called from `rules/no-cycle`.
We avoid doing unnecessary work by `for` being memoized (leveraging caching), and when we're done with the dependency graph we need these sets immediately to run the SCC algorithm.